xayn-ai/src/reranker/sync.rs

```diff
@@ -51,41 +50,21 @@ impl SyncData {
 
         reduce_cois(&mut self.user_interests.positive);
         reduce_cois(&mut self.user_interests.negative);
-
-        self.user_interests.reassign_ids();
     }
 }
 
 impl UserInterests {
     /// Moves all user interests of `other` into `Self`.
     pub(crate) fn append(&mut self, mut other: Self) {
-        // TODO drop dupes in other
         append_cois(&mut self.positive, &mut other.positive);
         append_cois(&mut self.negative, &mut other.negative);
     }
-
-    /// Re-assigns CoI ids for normalization.
-    pub(crate) fn reassign_ids(&mut self) {
-        reassign_coi_ids(&mut self.positive);
-        reassign_coi_ids(&mut self.negative);
-    }
 }
 
-fn append_cois<C>(locals: &mut Vec<C>, remotes: &mut Vec<C>)
-where
-    C: CoiPoint,
-{
-    // shift remote ids to avoid clashes with local ids
-    let max_local_id = locals.iter().map(|coi| coi.id().0).max().unwrap_or(0);
-    remotes
-        .iter_mut()
-        .for_each(|coi| coi.set_id(max_local_id + coi.id().0));
-
+/// Appends `remotes` to `locals`.
+///
+/// Remote CoIs with ids clashing with any of the local CoIs are removed.
+fn append_cois<C: CoiPoint>(locals: &mut Vec<C>, remotes: &mut Vec<C>) {
+    remotes.retain(|rem| !locals.iter().any(|loc| loc.id() == rem.id()));
```

Comment:
Do I assume correctly that we remove colliding Cois because we assume matches means that they had been synced before? (Assuming that UUID collisions in general don't happen.)

Besides that how many ids will this have in general. If they are many, we might want to use a "append to hash-set" schema instead of an retain + iter schema which is O(n^2).

Reply:
yes, it means they have been synced before and neither has been updated since the last sync, assuming uuid collisions are astronomically unlikely.

how often this happens would depend on usage I suppose, e.g. if there's little activity between syncs, not many CoIs will be updated, the more likely there will be dupes like this. using something like a hash set would be more optimal in general yes - this current version was just to get something simple working.